File: backend/src/main/java/com/paf/chop/backend/services/LearningPlanService.java
package com.paf.chop.backend.services;

import com.paf.chop.backend.dto.request.LearningPlanRequestDTO;
import com.paf.chop.backend.dto.request.ResourceCompletionDTO;
import com.paf.chop.backend.dto.response.LearningPlanResponseDTO;

import java.util.List;

public interface LearningPlanService {
    
    /**
     * Creates a new learning plan for a user
     * 
     * @param requestDTO the learning plan details
     * @param userId the ID of the user creating the plan
     * @return the created learning plan response
     */
    LearningPlanResponseDTO createLearningPlan(LearningPlanRequestDTO requestDTO, Long userId);
    
    /**
     * Updates an existing learning plan
     * 
     * @param id the ID of the learning plan to update
     * @param requestDTO the updated learning plan details
     * @param userId the ID of the user updating the plan
     * @return the updated learning plan response
     */
    LearningPlanResponseDTO updateLearningPlan(Long id, LearningPlanRequestDTO requestDTO, Long userId);
    
    /**
     * Deletes a learning plan
     * 
     * @param id the ID of the learning plan to delete
     * @param userId the ID of the user deleting the plan
     */
    void deleteLearningPlan(Long id, Long userId);
    
    /**
     * Retrieves a learning plan by ID
     * 
     * @param id the ID of the learning plan to retrieve
     * @param userId the ID of the user requesting the plan
     * @return the learning plan response
     */
    LearningPlanResponseDTO getLearningPlanById(Long id, Long userId);
    
    /**
     * Retrieves all learning plans for a user
     * 
     * @param userId the ID of the user
     * @return list of learning plan responses
     */
    List<LearningPlanResponseDTO> getUserLearningPlans(Long userId);
    
    /**
     * Retrieves all public learning plans
     * 
     * @return list of public learning plan responses
     */
    List<LearningPlanResponseDTO> getPublicLearningPlans();
    
    /**
     * Updates the completion status of a resource in a learning plan
     * 
     * @param planId the ID of the learning plan
     * @param completionDTO the resource completion details
     * @param userId the ID of the user updating completion status
     * @return the updated learning plan response
     */
    LearningPlanResponseDTO updateResourceCompletion(Long planId, ResourceCompletionDTO completionDTO, Long userId);
    
    /**
     * Calculates the progress percentage for a learning plan
     * 
     * @param planId the ID of the learning plan
     * @return the progress percentage (0-100)
     */
    Integer calculateProgress(Long planId);

    /**
     * Searches for public learning plans by keyword in title
     * 
     * @param keyword the search keyword
     * @return list of matching public learning plan responses
     */
    List<LearningPlanResponseDTO> searchPublicLearningPlans(String keyword);
}

File: frontend/src/services/learningPlanService.ts
import api from '../utils/axiosConfig';
import { LearningPlan, LearningPlanFormData } from '../types/learningPlan';

export const learningPlanService = {
    // Get all learning plans for the current user
    getUserLearningPlans: async (): Promise<LearningPlan[]> => {
        const response = await api.get('/learning-plans/me');
        return response.data;
    },

    // Get all public learning plans
    getPublicLearningPlans: async (): Promise<LearningPlan[]> => {
        const response = await api.get('/learning-plans/public');
        return response.data;
    },

    // Get a specific learning plan by ID
    getLearningPlanById: async (id: number): Promise<LearningPlan> => {
        const response = await api.get(`/learning-plans/${id}`);
        return response.data;
    },

    // Get a public learning plan by ID
    getPublicLearningPlanById: async (id: number): Promise<LearningPlan> => {
        const response = await api.get(`/learning-plans/public/${id}`);
        return response.data;
    },

    // Create a new learning plan
    createLearningPlan: async (planData: LearningPlanFormData): Promise<LearningPlan> => {
        const response = await api.post('/learning-plans', planData);
        return response.data;
    },

    // Update an existing learning plan
    updateLearningPlan: async (id: number, planData: LearningPlanFormData): Promise<LearningPlan> => {
        const response = await api.put(`/learning-plans/${id}`, planData);
        return response.data;
    },

    // Delete a learning plan
    deleteLearningPlan: async (id: number): Promise<void> => {
        await api.delete(`/learning-plans/${id}`);
    },

    // Search public learning plans
    searchPublicLearningPlans: async (keyword: string): Promise<LearningPlan[]> => {
        const response = await api.get(`/learning-plans/search?keyword=${encodeURIComponent(keyword)}`);
        return response.data;
    },

    // Mark a resource as completed or not completed
    updateResourceCompletion: async (planId: number, resourceId: number, completed: boolean): Promise<LearningPlan> => {
        const response = await api.post(
            `/learning-plans/${planId}/resources/${resourceId}/complete?completed=${completed}`
        );
        return response.data;
    }
}; 